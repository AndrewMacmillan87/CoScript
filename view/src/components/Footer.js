import React, { Fragment } from "react";
import { Link, withRouter } from "react-router-dom";

// Redux
import { connect } from "react-redux";
import { getMostRecentProject } from "../redux/actions/documents";

import PropTypes from "prop-types";

import "./footer.css";

export const Footer = ({ user, loading, getMostRecentProject, ...props }) => {
    const authLinks = (
        <div className='footer-links'>
            <Link to='/'>Home</Link>
            <Link to='/profile'>Profile</Link>
            <Link
                to='#'
                onClick={async () => {
                    await getMostRecentProject();
                    const { projectData } = props;
                    projectData
                        ? props.history.push(`/editor${projectData.url}`)
                        : props.history.push("/editor");
                }}>
                Workspace
            </Link>
            <Link to='/projects'>Active Projects</Link>
            <Link to='/publications'>Publications</Link>
        </div>
    );

    const guestLinks = (
        <div className='footer-links'>
            <Link to='/'>Home</Link>
            <Link to='/projects'>Active Projects</Link>
            <Link to='/publications'>Publications</Link>
        </div>
    );

    const isLoggedIn = user ? user.isLoggedIn : false;

    return (
        <footer className='footer'>
            <h2 className='footer-h2'>CoScript</h2>
            <div className='footer-container'>
                <div className='links-container socials'>
                    <p>Socials</p>
                    <div className='footer-links'>
                        <a href='https://github.com/AndrewMacmillan87/coscript'>
                            GitHub
                        </a>
                        <a href='/'>Twitter</a>
                        <a href='/'>Facebook</a>
                    </div>
                </div>
                <div className='links-container site-map'>
                    <p>Site Map</p>

                    {!loading && (
                        <Fragment>
                            {isLoggedIn ? authLinks : guestLinks}
                        </Fragment>
                    )}
                </div>
                <div className='links-container legals'>
                    <p>Legal</p>
                    <div className='footer-links'>
                        <a href='/terms'>Terms and Conditions</a>
                        <a href='/copyright'>Copyright Info</a>
                        <a href='/agreement'>Writers Agreements</a>
                        <a href='/'>Accreditation</a>
                    </div>
                </div>
            </div>
        </footer>
    );
};

Footer.propTypes = {
    user: PropTypes.object,
    loading: PropTypes.bool,
    getMostRecentProject: PropTypes.func.isRequired,
};

const mapProps = (state) => ({
    user: state.auth.user,
    loading: state.auth.loading,
    projectData: state.documents.projectData,
});

export default connect(mapProps, { getMostRecentProject })(withRouter(Footer));
