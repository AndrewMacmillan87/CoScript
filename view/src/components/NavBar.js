import React, { Fragment } from "react";
import { Link, withRouter } from "react-router-dom";
import Button from "./buttons/Button";

import { getUserMessages } from "../redux/auxiliary/user-messages";

// Redux
import { connect } from "react-redux";
import { getMostRecentProject } from "../redux/actions/documents";

import PropTypes from "prop-types";

import "./navbar.css";

export class NavBar extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            visible: false,
            unread: false,
        };
        this.handleClick = this.handleClick.bind(this);
    }

    /**
     * Checks to see if the user has any unread messages
     * in their personal profile message inbox and sets
     * the 'unread' state property to true if they have.
     * This displays a small icon to the user to indicate
     * they have unread messages
     */
    async componentDidMount() {
        const userMessages = await getUserMessages();
        for (let message of userMessages) {
            if (!message.read) {
                this.setState({
                    unread: true,
                });
            }
        }
    }

    /**
     * Sets the 'visible' state property so that the
     * dropdown menu can be revealed and hidden by
     * clicking the link
     */
    handleClick() {
        this.setState((state) => ({
            visible: !state.visible,
        }));
    }

    render() {
        const { unread } = this.state;
        const { user, loading, history } = this.props;

        const menuIcon = require("../img/menu.svg");
        const spinner = require("../img/loading.svg");
        const isLoggedIn = user && user.isLoggedIn | false;

        const loggedIn = (
            <Fragment>
                <div className='header'>
                    <div className='logo-container'>
                        <Link to='/'>
                            <p className='logo'>CoScript</p>
                        </Link>
                    </div>
                    <div className='nav-links'>
                        <Link to='/projects'>Active Projects</Link>
                        <Link to='/publications'>Publications</Link>
                        <Link
                            to='#'
                            onClick={async () => {
                                await this.props.getMostRecentProject();
                                const { projectData } = this.props;
                                projectData
                                    ? history.push(`/editor${projectData.url}`)
                                    : history.push("/editor");
                            }}>
                            Workspace
                        </Link>
                    </div>
                </div>
                <div className='nav-btns'>
                    {user && !loading ? (
                        <div className='logged-in-user-container'>
                            <p className='user-username'>
                                {user.profile.displayName}
                            </p>
                            <button
                                className='desktop-icon-link'
                                onClick={this.handleClick}>
                                {unread ? (
                                    <div className='avatar-unread-msg-container'>
                                        <div className='avatar-unread-msg-icon'></div>
                                        <img
                                            className='user-gravatar'
                                            src={user.avatar}
                                            alt='user gravatar'
                                        />
                                    </div>
                                ) : (
                                    <img
                                        className='user-gravatar'
                                        src={user.avatar}
                                        alt='user gravatar'
                                    />
                                )}
                            </button>
                        </div>
                    ) : (
                        <img
                            className='user-spinner'
                            src={spinner}
                            alt='loading spinner icon'
                        />
                    )}
                </div>
                <button
                    className='dropdown-icon-link'
                    onClick={this.handleClick}>
                    <img
                        className='dropdown-icon'
                        src={menuIcon}
                        alt='Burger menu icon'
                    />
                </button>
                <div
                    id='dropdown-nav'
                    className={
                        this.state.visible
                            ? "dropdown-menu dropdown-content display"
                            : "dropdown-menu dropdown-content"
                    }>
                    <Link to='/projects'>
                        <p className='dropdown-link'>Active Projects</p>
                    </Link>
                    <Link to='/publications'>
                        <p className='dropdown-link'>Publications</p>
                    </Link>
                    <Link
                        to='#'
                        onClick={async () => {
                            await this.props.getMostRecentProject();
                            const { projectData } = this.props;
                            projectData
                                ? history.push(`/editor${projectData.url}`)
                                : history.push("/editor");
                        }}>
                        <p className='dropdown-link'>Workspace</p>
                    </Link>
                    <Link to='/profile'>
                        <p className='dropdown-link'>My Profile</p>
                    </Link>
                    <a
                        onClick={() => {
                            history["from"] = "logout";
                            history.push("/redirect");
                        }}
                        href=' '>
                        <p className='dropdown-link'>Logout</p>
                    </a>
                </div>
            </Fragment>
        );

        const visiting = (
            <Fragment>
                <div className='header'>
                    <div className='logo-container'>
                        <Link to='/'>
                            <p className='logo'>CoScript</p>
                        </Link>
                    </div>
                    <div className='nav-links'>
                        <Link to='/projects'>Active Projects</Link>
                        <Link to='/publications'>Publications</Link>
                    </div>
                </div>
                <div className='nav-btns'>
                    <div>
                        <Link to='/register'>
                            <Button classes='btn' value='Register' />
                        </Link>
                    </div>
                    <div>
                        <Link to='/login'>
                            <Button classes='grn-btn' value='Login' />
                        </Link>
                    </div>
                </div>
                <button
                    className='dropdown-icon-link'
                    onClick={this.handleClick}>
                    <img
                        className='dropdown-icon'
                        src={menuIcon}
                        alt='Burger menu icon'
                    />
                </button>
                <div
                    id='dropdown-nav'
                    className={
                        this.state.visible
                            ? "dropdown-menu dropdown-content display"
                            : "dropdown-menu dropdown-content"
                    }>
                    <Link to='/register'>
                        <p className='dropdown-link'>Register</p>
                    </Link>
                    <Link to='/login'>
                        <p className='dropdown-link'>Login</p>
                    </Link>
                    <Link to='/projects'>
                        <p className='dropdown-link'>Active Projects</p>
                    </Link>
                    <Link to='/publications'>
                        <p className='dropdown-link'>Publications</p>
                    </Link>
                </div>
            </Fragment>
        );

        return (
            <nav className='navbar'>
                {!loading && (
                    <Fragment>{!isLoggedIn ? visiting : loggedIn}</Fragment>
                )}
            </nav>
        );
    }
}

NavBar.propTypes = {
    user: PropTypes.object,
    loading: PropTypes.bool,
    getMostRecentProject: PropTypes.func.isRequired,
};

const setProps = (state) => ({
    user: state.auth.user,
    loading: state.auth.loading,
    projectData: state.documents.projectData,
});

export default connect(setProps, {
    getMostRecentProject,
})(withRouter(NavBar));
