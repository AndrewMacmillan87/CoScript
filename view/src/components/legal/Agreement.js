import React, { Fragment, useEffect } from "react";

import NavBar from "../NavBar";
import Footer from "../Footer";

import "./legal.css";

export default function Agreement() {
    useEffect(() => {
        document.title = "CoScript - Agreements";
        window.scrollTo(0, 0);
    }, []);

    const informationLink = require("../../img/agreement.PNG");
    const agreementLink = require("../../img/agreement2.PNG");

    return (
        <Fragment>
            <NavBar />
            <main className='agreement-main'>
                <header className='agreement-header'>
                    <h1 className='agreement-h1'>
                        Collaborative Writers Agreements
                    </h1>
                </header>
                <p className='agreement-introduction'>
                    This area provides information on how you can draft up a
                    Collaborative Writers Agreement, and the reasons why you may
                    want to do so. This may not be something that every user
                    feels is necessary, but for some users it could provide an
                    extra layer of security and comfort whilst working with
                    others. Please be considerate when working with other
                    people, and if you feel that a formal agreement will help
                    this process then follow the links to learn how to put an
                    agreement in place.
                </p>
                <div className='source-info-container'>
                    <h2 className='source-info-header'>Rights of Writers</h2>
                    <a
                        href='http://www.rightsofwriters.com/2011/01/good-fences-when-and-why-co-writers.html'
                        rel='noopener noreferrer'
                        target='_blank'>
                        <img
                            className='source-link-img'
                            src={agreementLink}
                            alt='wiki-how website landing page'
                        />
                    </a>
                    <p className='source-info-description'>
                        Legal advice on entering into a collaborative writing
                        agreement, including a few examples of where one has
                        been beneficial in settling disputes. Sets out a list of
                        considerations that authors may find useful when
                        drafting their own agreements{" "}
                        <a
                            href='http://www.rightsofwriters.com/2011/01/good-fences-when-and-why-co-writers.html'
                            rel='noopener noreferrer'
                            target='_blank'>
                            http://www.rightsofwriters.com/2011/01/good-fences-when-and-why-co-writers.html
                        </a>
                    </p>
                </div>
                <div className='source-info-container'>
                    <h2 className='source-info-header'>Rocket Lawyers</h2>
                    <a
                        href='https://www.rocketlawyer.com/gb/en/sem/collaboration-agreement'
                        rel='noopener noreferrer'
                        target='_blank'>
                        <img
                            className='source-link-img'
                            src={informationLink}
                            alt='wiki-how website landing page'
                        />
                    </a>
                    <p className='source-info-description'>
                        A free and simple online contract provider that allows
                        users to draft their own agreements and add as many
                        participants as they like. Many users will not want to
                        share their information and this removes the need to
                        share personal details like email addresses that would
                        be needed in order to sign a hard copy of an agreement{" "}
                        <a
                            href='https://www.rocketlawyer.com/gb/en/sem/collaboration-agreement'
                            rel='noopener noreferrer'
                            target='_blank'>
                            https://www.rocketlawyer.com/gb/en/sem/collaboration-agreement
                        </a>
                    </p>
                </div>
            </main>
            <Footer />
        </Fragment>
    );
}
