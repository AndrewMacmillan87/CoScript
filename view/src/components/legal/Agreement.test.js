import React from "react";
import ReactDOM from "react-dom";
import { MemoryRouter as Router } from "react-router-dom";
import renderer from "react-test-renderer";
import { mount } from "enzyme";

import { Provider } from "react-redux";
import { reduxStore } from "../../redux/store/reduxStore";

import Agreement from "./Agreement";
import NavBar from "../NavBar";
import Footer from "../Footer";

describe("Agreement component", () => {
    test("renders without crashing", () => {
        global.scrollTo = jest.fn();
        const div = document.createElement("div");
        ReactDOM.render(
            <Provider store={reduxStore}>
                <Router>
                    <Agreement />
                </Router>
            </Provider>,
            div
        );
        ReactDOM.unmountComponentAtNode(div);
    });

    it("renders correctly", () => {
        global.scrollTo = jest.fn();
        const tree = renderer
            .create(
                <Provider store={reduxStore}>
                    <Router>
                        <Agreement />
                    </Router>
                </Provider>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render the NavBar component", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Agreement />
                </Router>
            </Provider>
        );
        expect(wrapper.find(NavBar).length).toEqual(1);
    });

    it("should render 1 main element with a className of agreement-main", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Agreement />
                </Router>
            </Provider>
        );
        expect(wrapper.find("main").length).toEqual(1);
        expect(wrapper.find({ className: "agreement-main" }).length).toEqual(1);
    });

    it("should render 1 header element with a className of agreement-header", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Agreement />
                </Router>
            </Provider>
        );
        expect(wrapper.find("header").length).toEqual(1);
        expect(wrapper.find({ className: "agreement-header" }).length).toEqual(
            1
        );
    });

    it("should render 6 p elements with a className of agreement-main", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Agreement />
                </Router>
            </Provider>
        );
        expect(wrapper.find("p").length).toEqual(6);
    });

    it("should render 2 img elements with a className of agreement-main", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Agreement />
                </Router>
            </Provider>
        );
        expect(wrapper.find("img").length).toEqual(2);
    });

    it("should render 11 a elements (including NavBar) with a className of agreement-main", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Agreement />
                </Router>
            </Provider>
        );
        expect(wrapper.find("a").length).toEqual(11);
    });

    it("should render the Footer component", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Agreement />
                </Router>
            </Provider>
        );
        expect(wrapper.find(Footer).length).toEqual(1);
    });
});
