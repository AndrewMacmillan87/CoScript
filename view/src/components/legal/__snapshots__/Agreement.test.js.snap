// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Agreement component renders correctly 1`] = `
Array [
  <nav
    className="navbar"
  />,
  <main
    className="agreement-main"
  >
    <header
      className="agreement-header"
    >
      <h1
        className="agreement-h1"
      >
        Collaborative Writers Agreements
      </h1>
    </header>
    <p
      className="agreement-introduction"
    >
      This area provides information on how you can draft up a Collaborative Writers Agreement, and the reasons why you may want to do so. This may not be something that every user feels is necessary, but for some users it could provide an extra layer of security and comfort whilst working with others. Please be considerate when working with other people, and if you feel that a formal agreement will help this process then follow the links to learn how to put an agreement in place.
    </p>
    <div
      className="source-info-container"
    >
      <h2
        className="source-info-header"
      >
        Rights of Writers
      </h2>
      <a
        href="http://www.rightsofwriters.com/2011/01/good-fences-when-and-why-co-writers.html"
        rel="noopener noreferrer"
        target="_blank"
      >
        <img
          alt="wiki-how website landing page"
          className="source-link-img"
          src="agreement2.PNG"
        />
      </a>
      <p
        className="source-info-description"
      >
        Legal advice on entering into a collaborative writing agreement, including a few examples of where one has been beneficial in settling disputes. Sets out a list of considerations that authors may find useful when drafting their own agreements
         
        <a
          href="http://www.rightsofwriters.com/2011/01/good-fences-when-and-why-co-writers.html"
          rel="noopener noreferrer"
          target="_blank"
        >
          http://www.rightsofwriters.com/2011/01/good-fences-when-and-why-co-writers.html
        </a>
      </p>
    </div>
    <div
      className="source-info-container"
    >
      <h2
        className="source-info-header"
      >
        Rocket Lawyers
      </h2>
      <a
        href="https://www.rocketlawyer.com/gb/en/sem/collaboration-agreement"
        rel="noopener noreferrer"
        target="_blank"
      >
        <img
          alt="wiki-how website landing page"
          className="source-link-img"
          src="agreement.PNG"
        />
      </a>
      <p
        className="source-info-description"
      >
        A free and simple online contract provider that allows users to draft their own agreements and add as many participants as they like. Many users will not want to share their information and this removes the need to share personal details like email addresses that would be needed in order to sign a hard copy of an agreement
         
        <a
          href="https://www.rocketlawyer.com/gb/en/sem/collaboration-agreement"
          rel="noopener noreferrer"
          target="_blank"
        >
          https://www.rocketlawyer.com/gb/en/sem/collaboration-agreement
        </a>
      </p>
    </div>
  </main>,
  <footer
    className="footer"
  >
    <h2
      className="footer-h2"
    >
      CoScript
    </h2>
    <div
      className="footer-container"
    >
      <div
        className="links-container socials"
      >
        <p>
          Socials
        </p>
        <div
          className="footer-links"
        >
          <a
            href="https://github.com/AndrewMacmillan87/coscript"
          >
            GitHub
          </a>
          <a
            href="/"
          >
            Twitter
          </a>
          <a
            href="/"
          >
            Facebook
          </a>
        </div>
      </div>
      <div
        className="links-container site-map"
      >
        <p>
          Site Map
        </p>
      </div>
      <div
        className="links-container legals"
      >
        <p>
          Legal
        </p>
        <div
          className="footer-links"
        >
          <a
            href="/terms"
          >
            Terms and Conditions
          </a>
          <a
            href="/copyright"
          >
            Copyright Info
          </a>
          <a
            href="/agreement"
          >
            Writers Agreements
          </a>
          <a
            href="/"
          >
            Accreditation
          </a>
        </div>
      </div>
    </div>
  </footer>,
]
`;
